### PR TITLE
feat(expo): Add RNSentrySDK APIs support to @sentry/react-native/expo plugin

### PR DESCRIPTION
**Pull request by antonis**
## :loudspeaker: Type of change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring

~~**Based on [feat: Capture app start errors before JS](https://github.com/getsentry/sentry-react-native/pull/4472/)**~~
**Based on https://github.com/getsentry/sentry-react-native/pull/5470**

## :scroll: Description

Adds RNSentrySDK APIs support to @sentry/react-native/expo plugin by importing sentry and adding injecting `RNSentrySDK.init/start` in the Android MainApplication (Kotlin or Java) or AppDelegate (Objective-C or Swift). 

  This feature is opt-out to enable it set `useNativeInit` to `true` in your `@sentry/react-native/expo` plugin configuration.

  ```js
  "plugins": [
    [
      "@sentry/react-native/expo",
      {
        "useNativeInit": true
      }
    ],
  ```

## :bulb: Motivation and Context

Fixes https://github.com/getsentry/sentry-react-native/issues/4625

## :green_heart: How did you test it?
CI, Manual

## :pencil: Checklist

- [x] I added tests to verify changes
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled
- [ ] I updated the docs if needed.
- [ ] I updated the wizard if needed.
- [x] All tests passing
- [x] No breaking changes

## :crystal_ball: Next steps

#skip-changelog

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,20 @@
 
 ### Features
 
+- Add RNSentrySDK APIs support to @sentry/react-native/expo plugin ([#4633](https://github.com/getsentry/sentry-react-native/pull/4633))
+
+  This feature is opt-out to enable it set `useNativeInit` to `true` in your `@sentry/react-native/expo` plugin configuration.
```
Comment (lucas-zimerman) [inline CHANGELOG.md:15]: NIT: It may make it easier for reading.
```suggestion
  This feature is opt-out, to enable it set `useNativeInit` to `true` in your `@sentry/react-native/expo` plugin configuration.

File: packages/core/plugin/src/withSentryAndroid.ts [hunk 2]
```diff
@@ -49,3 +56,50 @@ export function modifyAppBuildGradle(buildGradle: string): string {
...
+        // Insert import statement after package declaration
+        config.modResults.contents = config.modResults.contents.replace(
+          /(package .*\n\n?)/,
+          `$1import io.sentry.react.RNSentrySDK\n`,
+        );
+      }
+      // Add RNSentrySDK.init
+      config.modResults.contents = config.modResults.contents.replace(
+        /(super\.onCreate\(\)[;\n]*)([ \t]*)/,
+        `$1\n$2RNSentrySDK.init(this)\n$2`,
```
Comment (lucas-zimerman) [inline packages/core/plugin/src/withSentryAndroid.ts:99]: q: I think we should warn the user if `RNSentrySDK.init` wasn't inserted, to avoid giving the impression the SDK was set in some edge case where it wasn't

**Review by lucas-zimerman — COMMENTED** [review]
Overall the PR looks good and thank you for the tests!
I still think we could fine tune a bit more the android part and after it we could get ready for merge

File: packages/core/plugin/src/withSentryAndroid.ts [hunk 3]
```diff
@@ -49,3 +56,50 @@ export function modifyAppBuildGradle(buildGradle: string): string {
...
+      if (!config.modResults.contents.includes('import io.sentry.react.RNSentrySDK')) {
+        // Insert import statement after package declaration
+        config.modResults.contents = config.modResults.contents.replace(
+          /(package .*\n\n?)/,
+          `$1import io.sentry.react.RNSentrySDK\n`,
+        );
+      }
+      // Add RNSentrySDK.init
+      config.modResults.contents = config.modResults.contents.replace(
+        /(super\.onCreate\(\)[;\n]*)([ \t]*)/,
```
[the remark this answers is not in the document]
Comment (antonis) [reply packages/core/plugin/src/withSentryAndroid.ts:98]: Good question @lucas-zimerman 👍 I've started implementing with parameters but then realised that the [Application.onCreate](https://developer.android.com/reference/android/app/Application#onCreate()) doesn't have parameters (the [Activity.onCreate](https://developer.android.com/guide/components/activities/activity-lifecycle#oncreate) does)

File: packages/core/plugin/src/withSentryAndroid.ts (hunk 2 above)
Comment (antonis) [reply packages/core/plugin/src/withSentryAndroid.ts:104]: Good catch @lucas-zimerman 🙇 
I've added warning for both platforms with d25db305aaed11e2a8d5e9657ffae4fcc5287eee

**Comment by antonis** [comment]
_Note: The [Build & Test / Type Check Typescript 3.8 (pull_request)](https://github.com/getsentry/sentry-react-native/actions/runs/14262614941/job/39977478017?pr=4633) failure should be fixed when we merge https://github.com/getsentry/sentry-react-native/pull/4673 from `main`_

File: CHANGELOG.md [hunk 4]
```diff
@@ -10,6 +10,20 @@
 
 ### Features
 
+- Add RNSentrySDK APIs support to @sentry/react-native/expo plugin ([#4633](https://github.com/getsentry/sentry-react-native/pull/4633))
+
+  This feature is opt-out, to enable it set `useNativeInit` to `true` in your `@sentry/react-native/expo` plugin configuration.
```
Comment (krystofwoldrich) [inline CHANGELOG.md:15]: What's the reason for making this opt-out?

This could break user apps if they don't have the sentry options JSON in their project; the added init would cause a runtime error.  

I think this should be opt-in first.

Comment (antonis) [reply CHANGELOG.md:15]: @krystofwoldrich You are right! I think I've misread the [issue](https://github.com/getsentry/sentry-react-native/issues/4625) 😓

Comment (antonis) [reply CHANGELOG.md]: Updated with adc81a54d475203b1c69e92855a5e7631339cb5f. I also removed the note from the changelog since it is opt-in first. I can describe how to opt-out on a docs PR.

File: packages/core/plugin/src/withSentryAndroid.ts [hunk 5]
```diff
@@ -49,3 +56,58 @@ export function modifyAppBuildGradle(buildGradle: string): string {
 
   return buildGradle.replace(pattern, match => `${applyFrom}\n\n${match}`);
 }
+
+export function modifyMainApplication(config: ExpoConfig): ExpoConfig {
+  return withMainApplication(config, async config => {
+    if (!config.modResults || !config.modResults.path) {
+      warnOnce('Skipping MainApplication modification because the file does not exist.');
```
Comment (krystofwoldrich) [inline packages/core/plugin/src/withSentryAndroid.ts:63]: Let's try to be more explicit about what this means for the user -> the native Android init won't be added.

````suggestion
      warnOnce("Can't add `RNSentrySDK.init` to Android MainApplication, because the file was not found.");
````

File: packages/core/plugin/src/withSentryAndroid.ts [hunk 6]
```diff
@@ -49,3 +56,58 @@ export function modifyAppBuildGradle(buildGradle: string): string {
...
+      return config;
+    }
+
+    const fileName = config.modResults.path.split('/').pop();
+
+    if (config.modResults.contents.includes('RNSentrySDK.init')) {
+      warnOnce(`Your '${fileName}' already contains 'RNSentrySDK.init'.`);
```
Comment (krystofwoldrich) [inline packages/core/plugin/src/withSentryAndroid.ts:70]:
````suggestion
      warnOnce(`Your '${fileName}' already contains 'RNSentrySDK.init', the native code won't be updated.`);
````

File: packages/core/plugin/src/withSentryAndroid.ts [hunk 7]
```diff
@@ -49,3 +56,58 @@ export function modifyAppBuildGradle(buildGradle: string): string {
...
 }
+
+export function modifyMainApplication(config: ExpoConfig): ExpoConfig {
+  return withMainApplication(config, async config => {
+    if (!config.modResults || !config.modResults.path) {
+      warnOnce('Skipping MainApplication modification because the file does not exist.');
+      return config;
+    }
+
+    const fileName = config.modResults.path.split('/').pop();
```
[the remark this answers is not in the document]
Comment (antonis) [reply packages/core/plugin/src/withSentryAndroid.ts]: Good idea 👍 
Updated with 5f4f7c59856acb85d0261fe859acfdf7debc2504

File: packages/core/plugin/src/withSentryAndroid.ts [hunk 8]
```diff
@@ -49,3 +56,58 @@ export function modifyAppBuildGradle(buildGradle: string): string {
...
+      // Add RNSentrySDK.init
+      const originalContents = config.modResults.contents;
+      config.modResults.contents = config.modResults.contents.replace(
+        /(super\.onCreate\(\)[;\n]*)([ \t]*)/,
+        `$1\n$2RNSentrySDK.init(this);\n$2`,
+      );
+      if (config.modResults.contents === originalContents) {
+        warnOnce(`Failed to insert 'RNSentrySDK.init'.`);
+      }
+    } else {
```
Comment (krystofwoldrich) [inline packages/core/plugin/src/withSentryAndroid.ts:91]: Can we explicitly check for `lang == kotlin`? It will make the script more resilient

File: packages/core/plugin/src/withSentryAndroid.ts [hunk 9]
```diff
@@ -49,3 +56,58 @@ export function modifyAppBuildGradle(buildGradle: string): string {
...
+        );
+      }
+      // Add RNSentrySDK.init
+      const originalContents = config.modResults.contents;
+      config.modResults.contents = config.modResults.contents.replace(
+        /(super\.onCreate\(\)[;\n]*)([ \t]*)/,
+        `$1\n$2RNSentrySDK.init(this);\n$2`,
+      );
+      if (config.modResults.contents === originalContents) {
+        warnOnce(`Failed to insert 'RNSentrySDK.init'.`);
```
Comment (krystofwoldrich) [inline packages/core/plugin/src/withSentryAndroid.ts:89]: Let's also log the file name.

File: packages/core/plugin/src/withSentryAndroid.ts [hunk 10]
```diff
@@ -49,3 +56,58 @@ export function modifyAppBuildGradle(buildGradle: string): string {
...
+    }
+
+    if (config.modResults.language === 'java') {
+      if (!config.modResults.contents.includes('import io.sentry.react.RNSentrySDK;')) {
+        // Insert import statement after package declaration
+        config.modResults.contents = config.modResults.contents.replace(
+          /(package .*;\n\n?)/,
+          `$1import io.sentry.react.RNSentrySDK;\n`,
+        );
+      }
```
Comment (krystofwoldrich) [inline packages/core/plugin/src/withSentryAndroid.ts:81]: Let's do this only if the init call added to avoid used code.

File: packages/core/plugin/src/withSentryAndroid.ts [hunk 11]
```diff
@@ -49,3 +56,58 @@ export function modifyAppBuildGradle(buildGradle: string): string {
...
+      config.modResults.contents = config.modResults.contents.replace(
+        /(super\.onCreate\(\)[;\n]*)([ \t]*)/,
+        `$1\n$2RNSentrySDK.init(this);\n$2`,
+      );
+      if (config.modResults.contents === originalContents) {
+        warnOnce(`Failed to insert 'RNSentrySDK.init'.`);
+      }
+    } else {
+      // Kotlin
+      if (!config.modResults.contents.includes('import io.sentry.react.RNSentrySDK')) {
```
Comment (krystofwoldrich) [inline packages/core/plugin/src/withSentryAndroid.ts:93]: Let do this after init added same as Java comment.

File: packages/core/plugin/src/withSentryAndroid.ts [hunk 12]
```diff
@@ -49,3 +56,58 @@ export function modifyAppBuildGradle(buildGradle: string): string {
...
+        );
+      }
+      // Add RNSentrySDK.init
+      const originalContents = config.modResults.contents;
+      config.modResults.contents = config.modResults.contents.replace(
+        /(super\.onCreate\(\)[;\n]*)([ \t]*)/,
+        `$1\n$2RNSentrySDK.init(this)\n$2`,
+      );
+      if (config.modResults.contents === originalContents) {
+        warnOnce(`Failed to insert 'RNSentrySDK.init'.`);
```
Comment (krystofwoldrich) [inline packages/core/plugin/src/withSentryAndroid.ts:107]: Let's add the file name to the log.

File: packages/core/plugin/src/withSentryIOS.ts [hunk 13]
```diff
@@ -79,3 +85,58 @@ export function addSentryWithBundledScriptsToBundleShellScript(script: string):
...
+      const originalContents = config.modResults.contents;
+      config.modResults.contents = config.modResults.contents.replace(
+        /(func application\([^)]*\) -> Bool \{)/s,
+        `$1\n    RNSentrySDK.start()`,
+      );
+      if (config.modResults.contents === originalContents) {
+        warnOnce(`Failed to insert 'RNSentrySDK.start()'.`);
+      }
+    } else {
+      // Objective-C
```
Comment (krystofwoldrich) [inline packages/core/plugin/src/withSentryIOS.ts:117]: Can we explicitly check for obj-c?

File: packages/core/plugin/src/withSentryIOS.ts [hunk 14]
```diff
@@ -79,3 +85,58 @@ export function addSentryWithBundledScriptsToBundleShellScript(script: string):
     (match: string) => `/bin/sh ${SENTRY_REACT_NATIVE_XCODE_PATH} ${match}`,
   );
 }
+
+export function modifyAppDelegate(config: ExpoConfig): ExpoConfig {
+  return withAppDelegate(config, async config => {
+    if (!config.modResults || !config.modResults.path) {
+      warnOnce('Skipping AppDelegate modification because the file does not exist.');
```
Comment (krystofwoldrich) [inline packages/core/plugin/src/withSentryIOS.ts:92]: Same as for Android let's state that the 'RNSentrySDK.start()' won't be added to the AppDelegate.

File: packages/core/plugin/src/withSentryIOS.ts [hunk 15]
```diff
@@ -79,3 +85,58 @@ export function addSentryWithBundledScriptsToBundleShellScript(script: string):
...
+    }
+
+    const fileName = path.basename(config.modResults.path);
+
+    if (config.modResults.language === 'swift') {
+      if (config.modResults.contents.includes('RNSentrySDK.start()')) {
+        warnOnce(`Your '${fileName}' already contains 'RNSentrySDK.start()'.`);
+        return config;
+      }
+      if (!config.modResults.contents.includes('import RNSentry')) {
```
Comment (krystofwoldrich) [inline packages/core/plugin/src/withSentryIOS.ts:103]: Let's add the imports after the start call is added.

File: packages/core/test/expo-plugin/modifyAppDelegate.test.ts [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+  }
+}`;
+
+describe('modifyAppDelegate', () => {
+  let config: MockedExpoConfig;
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+    // Reset to a mocked Swift config after each test
+    config = {
```
Comment (krystofwoldrich) [inline packages/core/test/expo-plugin/modifyAppDelegate.test.ts:98]: Let's deep copy or use func to create the config to avoid comparing the same reference int he test with no modification.

File: packages/core/test/expo-plugin/modifyMainApplication.test.ts [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+}
+`;
+
+describe('modifyMainApplication', () => {
+  let config: MockedExpoConfig;
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+    // Reset to a mocked Java config after each test
+    config = {
```
Comment (krystofwoldrich) [inline packages/core/test/expo-plugin/modifyMainApplication.test.ts:128]: Same as for iOS let's avoid comparing the same ref.

File: packages/core/test/expo-plugin/modifyAppDelegate.test.ts (hunk 16 above)
Comment (antonis) [reply packages/core/test/expo-plugin/modifyAppDelegate.test.ts]: Good catch 👍 Updated in 744993c8255ca160e41673f01cc09f7e58eb3c69

File: packages/core/plugin/src/withSentryIOS.ts (hunk 15 above)
Comment (antonis) [reply packages/core/plugin/src/withSentryIOS.ts]: Updated with 0ffd26c8f16ee2e1a1bb04bca13be7a95ff2726b

File: packages/core/plugin/src/withSentryIOS.ts (hunk 14 above)
Comment (antonis) [reply packages/core/plugin/src/withSentryIOS.ts]: Updated with dce74b25792f6d9b1b1da7e894642b365c9890ed

File: packages/core/plugin/src/withSentryIOS.ts (hunk 13 above)
Comment (antonis) [reply packages/core/plugin/src/withSentryIOS.ts]: Updated with a53c7f46a14361e8f3b2b1da7f52b4f404799751

File: packages/core/plugin/src/withSentryAndroid.ts (hunk 12 above)
Comment (antonis) [reply packages/core/plugin/src/withSentryAndroid.ts]: Updated with 235f3efbe6180b41d45a748b04fd4d95393873b7

File: packages/core/plugin/src/withSentryAndroid.ts (hunk 11 above)
Comment (antonis) [reply packages/core/plugin/src/withSentryAndroid.ts]: Updated with 369cce774d10f252e53399b708f7c3d6c6042190

File: packages/core/plugin/src/withSentryAndroid.ts (hunk 10 above)
Comment (antonis) [reply packages/core/plugin/src/withSentryAndroid.ts]: Good point 👍 Updated with 369cce774d10f252e53399b708f7c3d6c6042190

File: packages/core/plugin/src/withSentryAndroid.ts (hunk 9 above)
Comment (antonis) [reply packages/core/plugin/src/withSentryAndroid.ts]: Updated with 235f3efbe6180b41d45a748b04fd4d95393873b7

File: packages/core/plugin/src/withSentryAndroid.ts (hunk 8 above)
Comment (antonis) [reply packages/core/plugin/src/withSentryAndroid.ts]: Good idea 👍 Updated with 62d39ccb4f99530bd9ee3188cf4b1220c3d5ab4d

File: packages/core/plugin/src/withSentryIOS.ts [hunk 18]
```diff
@@ -79,3 +85,57 @@ export function addSentryWithBundledScriptsToBundleShellScript(script: string):
...
+        return config;
+      }
+      // Add RNSentrySDK.start() at the beginning of application method
+      const originalContents = config.modResults.contents;
+      config.modResults.contents = config.modResults.contents.replace(
+        /(func application\([^)]*\) -> Bool \{)/s,
+        `$1\n    RNSentrySDK.start()`,
+      );
+      if (config.modResults.contents === originalContents) {
+        warnOnce(`Failed to insert 'RNSentrySDK.start()' in '${fileName}.`);
```
Comment (lucas-zimerman) [inline packages/core/plugin/src/withSentryIOS.ts:110]: NIT: missing  '
```suggestion
        warnOnce(`Failed to insert 'RNSentrySDK.start()' in '${fileName}'.`);
```

**Review by lucas-zimerman — APPROVED** [review]
I have re-checked the PR and overall, I only found a nit on a string. After fixing it and if @krystofwoldrich doesn't find anything else, LGTM!

File: packages/core/plugin/src/withSentryIOS.ts (hunk 18 above)
Comment (antonis) [reply packages/core/plugin/src/withSentryIOS.ts:110]: Good catch 🙇

File: packages/core/plugin/src/withSentryAndroid.ts [hunk 19]
```diff
@@ -49,3 +56,57 @@ export function modifyAppBuildGradle(buildGradle: string): string {
 
   return buildGradle.replace(pattern, match => `${applyFrom}\n\n${match}`);
 }
+
+export function modifyMainApplication(config: ExpoConfig): ExpoConfig {
+  return withMainApplication(config, async config => {
```
Comment (krystofwoldrich) [inline packages/core/plugin/src/withSentryAndroid.ts]: There are no async operations in the function.

````suggestion
  return withMainApplication(config, config => {
````

File: packages/core/plugin/src/withSentryAndroid.ts [hunk 20]
```diff
@@ -49,3 +56,57 @@ export function modifyAppBuildGradle(buildGradle: string): string {
...
+      warnOnce(`Your '${fileName}' already contains 'RNSentrySDK.init', the native code won't be updated.`);
+      return config;
+    }
+
+    if (config.modResults.language === 'java') {
+      // Add RNSentrySDK.init
+      const originalContents = config.modResults.contents;
+      config.modResults.contents = config.modResults.contents.replace(
+        /(super\.onCreate\(\)[;\n]*)([ \t]*)/,
+        `$1\n$2RNSentrySDK.init(this);\n$2`,
```
Comment (krystofwoldrich) [inline packages/core/plugin/src/withSentryAndroid.ts]: Nice use of the regex, I really like how it matches the existing code white space prefix.

File: packages/core/plugin/src/withSentryIOS.ts [hunk 21]
```diff
@@ -12,8 +13,11 @@ const SENTRY_REACT_NATIVE_XCODE_PATH =
 const SENTRY_REACT_NATIVE_XCODE_DEBUG_FILES_PATH =
   "`${NODE_BINARY:-node} --print \"require('path').dirname(require.resolve('@sentry/react-native/package.json')) + '/scripts/sentry-xcode-debug-files.sh'\"`";
 
-export const withSentryIOS: ConfigPlugin<string> = (config, sentryProperties: string) => {
-  const cfg = withXcodeProject(config, config => {
+export const withSentryIOS: ConfigPlugin<{ sentryProperties: string; useNativeInit: boolean | undefined }> = (
+  config,
+  { sentryProperties, useNativeInit = true },
```
Comment (krystofwoldrich) [inline packages/core/plugin/src/withSentryIOS.ts]: I believe `useNativeInit` must be `false` by default, otherwise it would break existing users.

By adding the native init the SDK would behave differently for user without the init file. The options would not be synced.

File: packages/core/plugin/src/withSentryIOS.ts [hunk 22]
```diff
@@ -79,3 +85,57 @@ export function addSentryWithBundledScriptsToBundleShellScript(script: string):
...
+    if (config.modResults.language === 'swift') {
+      if (config.modResults.contents.includes('RNSentrySDK.start()')) {
+        warnOnce(`Your '${fileName}' already contains 'RNSentrySDK.start()'.`);
+        return config;
+      }
+      // Add RNSentrySDK.start() at the beginning of application method
+      const originalContents = config.modResults.contents;
+      config.modResults.contents = config.modResults.contents.replace(
+        /(func application\([^)]*\) -> Bool \{)/s,
+        `$1\n    RNSentrySDK.start()`,
```
Comment (krystofwoldrich) [inline packages/core/plugin/src/withSentryIOS.ts]: Would it make sense to also use a second capture group to dynamically fill in the tabs/spaces?

File: packages/core/plugin/src/withSentryIOS.ts (hunk 21 above)
Comment (antonis) [reply packages/core/plugin/src/withSentryIOS.ts]: Good catch 👍 
Updated with c356288dbbb4afb1b82f897ecc720bf1e457b015

File: packages/core/plugin/src/withSentryIOS.ts (hunk 22 above)
Comment (antonis) [reply packages/core/plugin/src/withSentryIOS.ts]: Good idea. Updated with 8e32556b26c27a334e47b425356602b003240cb2

File: packages/core/plugin/src/withSentryIOS.ts [hunk 23]
```diff
@@ -79,3 +85,57 @@ export function addSentryWithBundledScriptsToBundleShellScript(script: string):
...
+        warnOnce(`Failed to insert '[RNSentrySDK start]' in '${fileName}.`);
+      } else if (!config.modResults.contents.includes('#import <RNSentry/RNSentry.h>')) {
+        // Add import after AppDelegate.h
+        config.modResults.contents = config.modResults.contents.replace(
+          /(#import "AppDelegate.h"\n)/,
+          `$1#import <RNSentry/RNSentry.h>\n`,
+        );
+      }
+    } else {
+      warnOnce(`Unsupported language detected in '${fileName}', the native code won't be updated.`);
```
Comment (krystofwoldrich) [inline packages/core/plugin/src/withSentryIOS.ts]: I've tried to run the plugin with our sample and got this message for iOS.

Might be worth to log out the value of the language.

File: packages/core/plugin/src/withSentryIOS.ts [hunk 24]
```diff
@@ -79,3 +85,57 @@ export function addSentryWithBundledScriptsToBundleShellScript(script: string):
...
+        /(func application\([^)]*\) -> Bool \{)\s*\n(\s*)/s,
+        `$1\n$2RNSentrySDK.start()\n$2`,
+      );
+      if (config.modResults.contents === originalContents) {
+        warnOnce(`Failed to insert 'RNSentrySDK.start()' in '${fileName}'.`);
+      } else if (!config.modResults.contents.includes('import RNSentry')) {
+        // Insert import statement after UIKit import
+        config.modResults.contents = config.modResults.contents.replace(/(import UIKit\n)/, `$1import RNSentry\n`);
+      }
+    } else if (config.modResults.language === 'objc') {
```
Comment (krystofwoldrich) [inline packages/core/plugin/src/withSentryIOS.ts]: Depending on RN version, we can encounter also `objcpp`.

```suggestion
    } else if (['objcpp', 'objc'].includes(config.modResults.language)) {
```

**Comment by krystofwoldrich** [comment]
I'm currently updating the base branch with the latest main since there was quite a lot of development that was not synced to the feature branch.

**Comment by antonis** [comment]
> I'm currently updating the base branch with the latest main since there was quite a lot of development that was not synced to the feature branch.

Thank you for the heads up @krystofwoldrich 🙇 
I'll update this branch after that along with the review feedback.

File: packages/core/plugin/src/withSentryIOS.ts (hunk 24 above)
Comment (antonis) [reply packages/core/plugin/src/withSentryIOS.ts]: Good idea 👍

File: packages/core/plugin/src/withSentryIOS.ts (hunk 23 above)
Comment (antonis) [reply packages/core/plugin/src/withSentryIOS.ts]: Good idea 👍 
Added the language in the warning for both iOS and Android with a20984c8fdc30deff9a04c8eac7dc9501404e81d

**Review by krystofwoldrich — APPROVED** [review]
Thank you for the fixes. 🚀 It looks great!

File: CHANGELOG.md [hunk 25]
```diff
@@ -10,6 +10,7 @@
 
 ### Features
 
+- Add RNSentrySDK APIs support to @sentry/react-native/expo plugin ([#4633](https://github.com/getsentry/sentry-react-native/pull/4633))
```
Comment (krystofwoldrich) [inline CHANGELOG.md:17]: Nit. It would be nice to include an example code snippet and a small summary of what will the flag do.

**Comment by lucas-zimerman** [comment]
@sentry review

**Comment by lucas-zimerman** [comment]
@antonis should we restart this PR?

**Comment by antonis** [comment]
> @antonis should we restart this PR?

I don't expect this to have major conflicts with main/v7.

File: packages/core/plugin/src/withSentryIOS.ts [hunk 26]
```diff
@@ -78,3 +85,59 @@ export function addSentryWithBundledScriptsToBundleShellScript(script: string):
...
+      if (config.modResults.contents.includes('RNSentrySDK.start()')) {
+        warnOnce(`Your '${fileName}' already contains 'RNSentrySDK.start()'.`);
+        return config;
+      }
+      // Add RNSentrySDK.start() at the beginning of application method
+      const originalContents = config.modResults.contents;
+      config.modResults.contents = config.modResults.contents.replace(
+        /(func application\([^)]*\) -> Bool \{)\s*\n(\s*)/s,
+        '$1\n$2RNSentrySDK.start()\n$2',
+      );
```
[the remark this answers is not in the document]
Comment (antonis) [reply packages/core/plugin/src/withSentryIOS.ts:108]: There IS a theoretical edge case where the regex could fail: if function parameters contain closing parentheses (like closure types):                                                                                                                                             
                                                                                                                                                                                                                                                                                    
  func application(_ app: UIApplication, handler: ((Int) -> Void)?) -> Bool {                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                                    
  Here, [^)]* would stop at the first ) in ((Int) -> Void), not the function's closing paren.                                                                                                                                                                                       
                                                                                                                                                                                                                                                                                    
  However: This is not a concern because:                                                                                                                                                                                                                                           
  - Expo-generated AppDelegates don't use such parameter types                                                                                                                                                                                                                      
  - The standard application(_:didFinishLaunchingWithOptions:) signature doesn't have nested parentheses                                                                                                                                                                            
  - All our test cases pass                                                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                                                    
 No fix needed.

**Comment by antonis** [comment]
@sentry review

File: packages/core/ios/RNSentry.h [hunk 27]
```diff
@@ -10,7 +10,7 @@
 #import <Sentry/Sentry.h>
 
 // This import exposes public RNSentrySDK start
-#import "RNSentrySDK.h"
+#import <RNSentry/RNSentrySDK.h>
```
Comment (lucas-zimerman) [inline packages/core/ios/RNSentry.h]: q: Why this changed?

**Comment by lucas-zimerman** [comment]
@sentry review

File: packages/core/plugin/src/logger.ts [hunk 28]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * Log a warning message only once per run.
+ * This is used to avoid spamming the console with the same message.
+ */
+export function warnOnce(message: string): void {
+  if (!warningMap.has(message)) {
+    warningMap.set(message, true);
+    // eslint-disable-next-line no-console
+    console.warn(yellow(prefix(message)));
```
Comment (lucas-zimerman) [inline packages/core/plugin/src/logger.ts:11]: q: should we temporarily disable the capture of console warnings as events here?

File: packages/core/plugin/src/version.ts [hunk 29]
```diff
@@ -0,0 +1,8 @@
+const packageJson: {
+  name: string;
+  version: string;
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
+} = require('../../package.json');
+
+export const PLUGIN_NAME = `${packageJson.name}/expo`;
+export const PLUGIN_VERSION = packageJson.version;
```
Comment (lucas-zimerman) [inline packages/core/plugin/src/version.ts:8]: shouldn't we protect these variables in case `packagejson` is not found?

File: packages/core/plugin/src/withSentry.ts [hunk 30]
```diff
@@ -79,6 +81,6 @@ ${authToken ? `${existingAuthTokenMessage}\nauth.token=${authToken}` : missingAu
 }
 
 // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
-const withSentry = createRunOncePlugin(withSentryPlugin, sdkPackage.name, sdkPackage.version);
+const withSentry = createRunOncePlugin(withSentryPlugin, PLUGIN_NAME, PLUGIN_VERSION);
```
Comment (lucas-zimerman) [inline packages/core/plugin/src/withSentry.ts:84]: might be a good idea to use the sdkPackage value as a fallback in case PLUGIN_NAME/VERSION are undefined.

**Comment by lucas-zimerman** [comment]
` I updated the docs if needed.` was marked and this is a feature. Do we already have a PR to document this?

**Comment by lucas-zimerman** [comment]
q: What will happen when an expo go app sets `useNativeInit: true`?

**Comment by antonis** [comment]
> ` I updated the docs if needed.` was marked and this is a feature. Do we already have a PR to document this?

Good catch @lucas-zimerman. I've unchecked those.

> q: What will happen when an expo go app sets `useNativeInit: true`?

My understanding is that it would fail similar to other features on runtime. The code injection would also fail with errors similar to [this](https://github.com/getsentry/sentry-react-native/pull/4633/changes#diff-e2bb8c37eebce5184fa1eefa69f6f8548aceab620e2409a2cdfc58a1c0ffc02dR63). I'll test more extensively this case though.

File: packages/core/ios/RNSentry.h (hunk 27 above)
Comment (antonis) [reply packages/core/ios/RNSentry.h]: My understanding is that the latest Swift changes on the iOS side require angle bracket imports

File: packages/core/plugin/src/logger.ts (hunk 28 above)
Comment (antonis) [reply packages/core/plugin/src/logger.ts:11]: I think this would appear during the build phase. Why should we disable it?

File: packages/core/plugin/src/version.ts (hunk 29 above)
Comment (antonis) [reply packages/core/plugin/src/version.ts:8]: If package.json doesn't exist, this import statement fails during build/compilation

File: packages/core/plugin/src/withSentry.ts (hunk 30 above)
Comment (antonis) [reply packages/core/plugin/src/withSentry.ts:84]: I think this should fail again at compile time if the import above fails.
